core/src/test/scala/com.mediative.sparrow/DataFrameReaderTest.scala

```diff
@@ -173,51 +195,107 @@ class DataFrameReaderTest extends FreeSpec with BeforeAndAfterAll {
+        val rdd1 = rdd1Maybe.toOption.get
+        assertResult(0) { rdd0.subtract(rdd1).count }
+        assertResult(0) { rdd1.subtract(rdd0).count }
+      }
 
-      assert(df.toRDD[TestToRdd1].toOption.get.first == expected)
-    }
+      import TestCaseClasses._
 
-    "round-trip an object containing an Int and an optional Int from RDD to DataFrame back to RDD" - {
-      "when containing some value" in {
-        import TestCaseClasses.TestToRdd2
+      "Int, String" in {
+        testRDD_DataFrame_codec(TestToRdd1(1, "a"))
+      }
 
-        val expected = TestToRdd2(1, Option(1))
-        val df =
-          sqlContext.createDataFrame(sc.parallelize(List(expected)))
+      "Int, Option[Int]" - {
+        "when Some(Int)" in {
+          testRDD_DataFrame_codec(TestToRdd2(1, Option(1)))
+        }
 
-        assert(df.toRDD[TestToRdd2].toOption.get.first == expected)
+        "when None" in {
+          testRDD_DataFrame_codec(TestToRdd2(1, Option.empty))
+        }
       }
 
-      "when containing no value" in {
-        import TestCaseClasses.TestToRdd2
-
-        val expected = TestToRdd2(1, Option.empty)
-        val df =
-          sqlContext.createDataFrame(sc.parallelize(List(expected)))
+      "String, java.sql.Timestamp" in {
+        pendingUntilFixed {
+          // FIXME:
+          // rdd1Maybe.isSuccess was false Failure(NonEmptyList(The field 'timestampVal' isn't a LongType as expected, TimestampType received.)) (DataFrameReaderTest.scala:203)
+          testRDD_DataFrame_codec(
+            TestToRdd3("a", Timestamp.valueOf("2015-07-15 09:00:00"))
+          )
+        }
+      }
 
-        assert(df.toRDD[TestToRdd2].toOption.get.first == expected)
+      "Int, Double" in {
+        pendingUntilFixed {
+          // FIXME:
+          // "org.apache.spark.SparkException: Job aborted due to stage failure"
+          // Caused by: java.lang.ClassCastException: java.lang.Double cannot be cast to java.lang.Integer
+          // at scala.runtime.BoxesRunTime.unboxToInt(BoxesRunTime.java:106)
+          testRDD_DataFrame_codec(TestToRdd4(1, 2.0))
+        }
       }
-    }
 
-    "round-trip an object containing a String and a Timestamp from RDD to DataFrame back to RDD" in {
-      import TestCaseClasses.TestToRdd3
+      "Int, Option[Double]" - {
+        "when Some(Double)" in {
+          testRDD_DataFrame_codec(TestToRdd5(1, Some(2.0)))
+        }
+        "when None" in {
+          testRDD_DataFrame_codec(TestToRdd5(1, None))
+        }
+      }
 
-      val expected =
-        TestToRdd3("a", Timestamp.valueOf("2015-07-15 09:00:00"))
+      "Int, Wrapped[String]" in {
+        pendingUntilFixed {
+          // FIXME:
+          // rdd1Maybe.isSuccess was false Failure(NonEmptyList(The field 'wrappedStringVal' isn't a StringType as expected, StructType(StructField(unwrap,StringType,true)) received.)) (DataFrameReaderTest.scala:207)
+          testRDD_DataFrame_codec(TestToRdd8(1, Wrap.Wrapped("foo")))
```

Comment:
@gpoirier i understand why it happens, and while it isn't a bug in existing code, it's definitely a gap in our existing functionality, imo.

for sparrow to be useful as a library the responsibility is on us to ensure that either this use case is supported (in some manner), or if it feasibly can't be, then for us to document what is the alternative to take. eg. specifically in this case:

- what would be the `RowConverter`/`FieldConverter` that you're suggesting should be used instead? is the suggestion to have a `FieldConverter` which says that if we're expecting a `Wrapped[T]` we instead get a `T`, that it's ok? if so, that hinders the ability of users to encode invariants/properties in types.

perhaps there isn't a better way than above (i don't know at this point), but either way our tests and README should make that explicit. if we find out that there isn't a better way, the `pendingUntilFixed` can be converted to an `intercept` call instead (and the README updated accordingly too to make clear that which we cannot support)